### DISCUSSION TITLE
how to write prepared statements in psycopg3

### DISCUSSION THREAD

**Discussion by EngineerHus** [psycopg/psycopg#492] [Q&A] [answered]
Are prepared statements written differently in psycopg3 than psycopg2?

I am upgrading from psycopg2 to 3 and in psycopg2 this works but doesnt work for psycopg3

Also the variables ```name``` and ```age``` will be populated with data

```sql``` is a variable which is connected to the Database we use. So my problem is within the query itself and not the python syntax
```
def test(name, age=None, alive=False):


    sql.execute("""prepare test as UPDATE table AS x
                SET var = $1
                FROM table1 as y
                WHERE y.name = x.name and y.age = $2""")
   sql.execute("""execute test (%s, %s)""", (name, age))
```

This throws an error saying ```could not determine the datatype of parameter $1```. Looking at the docs and github of psycopg, theres no info regarding this matter. Has prepared statements completely been reengineered?

**Comment by dvarrazzo** [comment, marked as answer]
Psycopg 3 [handles prepared statements automatically](https://www.psycopg.org/psycopg3/docs/advanced/prepare.html), psycopg2 doesn't handle them at all. Preparing a statement manually in pg3 that way doesn't make a lot of sense: just run the update, passing `prepare=True` if you wish, but if you don't, and the query is executed a few times, it will be prepared automatically.

The statement you prepare manually probably loses type info, because types are not specified [as postgres allows](https://www.postgresql.org/docs/current/sql-prepare.html).

The error, ultimately, seems to come from the fact that there is [no server type info on the first variable](https://www.psycopg.org/psycopg3/docs/basic/from_pg2.html#different-cast-rules), and psycopg doesn't give any, because the value is either None or a string (which is passed with unknown data type to the query). One of the following might solve it:

- leave psycopg do the statement preparation. It will be more efficient too.
- specify a `%s::text` or other type on your first query placeholder
- specify the types in your `PREPARE` statement

In order of preference.

**Reply by EngineerHus** [reply]
"pg3 handles prepared statements automatically", does this mean that prepared statements are prepared automatically behind the scenes where using the keyword PREPARE is useless now in pg3?

**Reply by dvarrazzo** [reply]
That's correct, see the documentation linked above

**Reply by EngineerHus** [reply]
thanks
